### ISSUE TITLE
[Bug] Crash on trying to cure a dead patient in Operating Theatre

### ISSUE THREAD

**Issue by Lobashev** [CorsixTH/CorsixTH#2562]
### Describe the issue

Surgeons will attempt to treat a dead patient. But they won't succeed.
There's a long line for the surgery room. I've built a second surgery room. The patient went to the second room and got in line.
![image](https://github.com/CorsixTH/CorsixTH/assets/89157487/ab2a0f98-68fc-45c2-9f88-be895706d3bb)

By the time it was his turn, he should have been dead.
![image](https://github.com/CorsixTH/CorsixTH/assets/89157487/453a3dac-3cf2-4a95-bc9d-20bb44453289)

But he went in. He was operated on. They didn't take any money. He started to change his clothes to get out. And then the game freezes.
![image](https://github.com/CorsixTH/CorsixTH/assets/89157487/765f30d3-b8d8-44fc-80e7-657a65b81a55)
![image](https://github.com/CorsixTH/CorsixTH/assets/89157487/da80d0f1-f79a-4546-8b1b-6229f858ffbc)

### Steps to Reproduce

1. Load Autosave9.sav
2. Move the camera to the right-bottom corner of the hospital, where the operating room is
3. Wait for the surgeons to treat the current patient
4. The next patient to come in is the dead patient. He will be operated on. After he gets up, they don't charge him. He will go to change his clothes and at that point the game will freeze

### Expected Behaviour

The patient should not go into the office, but die in the hallway

### Save Game

[Autosave9.zip](https://github.com/CorsixTH/CorsixTH/files/15058971/Autosave9.zip)

### CorsixTH Version

0.67

### Operating System

Windows 11

### Theme Hospital Version

Origin

### Gamelog.txt

```bash
Gamelog generated on 2024-04-22 10:10:44
Compiled with api_version: 2684, arch: x64, audio: true, os: windows
SDL renderer: direct3d
Lua 5.4 run with api version: 2684, game version: v0.67, savegame version: 180
Created game with savegame version 180.
Savegame version is 180 (v0.67), originally it was 180 (v0.67).

Error in timer handler: 
Lua\entities\humanoid.lua:472: assertion failed!
stack traceback:
	[C]: in function 'assert'
	Lua\entities\humanoid.lua:472: in method 'startAction'
	Lua\entities\humanoid.lua:577: in method 'finishAction'
	Lua\humanoid_actions\use_screen.lua:48: in local 'timer_function'
	Lua\entity.lua:216: in method 'tick'
	Lua\world.lua:1074: in method 'onTick'
	Lua\app.lua:1202: in function <Lua\app.lua:1199>
	(...tail calls...)
	Lua\app.lua:1102: in function <Lua\app.lua:1097>
```

**Comment by lewri** [comment]
Thanks for this report. Is the error reproducable from the save provided or has it already happened?

**Comment by lewri** [comment]
> By the time it was his turn, he should have been dead.

Did he get up from the bench then sit back down again do you know?

**Comment by Lobashev** [comment]
@lewri 
> Thanks for this report. Is the error reproducable from the save provided or has it already happened?

Can be reproduced from save


> Did he get up from the bench then sit back down again do you know?

Just got up and went to the office. This can be seen in the save

**Comment by lewri** [comment]
```lua
Clicked on:
humanoid - class: Standard Male Patient
Warmth: 0.324   Happiness: 1.000   Fatigue: 0.000  Thirst: 0.493  Toilet_Need: 0.528   Health: 0.000   Service Quality: 0.000
Actions: [use_object - bench   must_happen, queue - Bench distance: 0 Standing: false   must_happen, walk - going to 93:68   must_happen, meander, die]
```

Already dead in the save. @mugmuggy if you're around any ideas if this may have slipped past #2310 ?

**Comment by mugmuggy** [comment]
@lewri Can't be definitive, but I would say its 'interrupted' twice. The obvious issue for me is that the queue action in the action queue position is immediately outside the door and the walk is to in the room. The must happen and saved must happen are true on the walk action so they will enter and it can't be interrupted until after the first tile is walked.

As there are those multiple bits of code that are interrupting actions that cause it to be written and processed makes it harder or uglier to resolve this specific case.

**Comment by lewri** [comment]
Thanks muggy,

When I was playing around I found that if you found a way to initiate some kind of queue update or kicked them off the bench the problematic must happen was unblocked. 

Should we still try and prevent a crash when it falls over though for now? It probably isn't always good enough to ignore an empty action queue if the going home flag is true.

**Comment by lewri** [comment]
> The obvious issue for me is that the queue action in the action queue position is immediately outside the door and the walk is to in the room. The must happen and saved must happen are true on the walk action so they will enter and it can't be interrupted until after the first tile is walked.

Replicated this, can confirm it's the bench position.

**Comment by ARGAMX** [comment]
The same crash happens in #2947 (`Autosave4.sav`)

[2947-Autosave4.sav.zip](https://github.com/user-attachments/files/21531355/2947-Autosave4.sav.zip)

**Comment by ARGAMX** [comment]
The game crashes if logic makes an attempt to cure a patient whose life timer has already expired and he should have died earlier.

Why didn't the patient die but ended up in the treatment room?

The problem is that the patient cannot start the dying scenario if he:
1. knocks on the door or opens it.
2. is in the room or in the toilet.
3. there may be some other cases that I didn't take into account.

instead, the dying scenario is simply cancels... and the patient start to live forever and moves on!

i.e. the logic of trying to start the dying scenario does not check whether this scenario can be executed or not. 
It just tries to run it once and that's it. one-time. If the run attempt fails, the patient lives on with the expired timer.

Obviously, to prevent this from happening, we need to:
1. run the dying scenario only if the patient can actually execute die this scenario now. if not, then postpone try for the next tick.

thus, theoretically, patients will stop getting into this invalid state.

1. this is the problem encountered by saves in 
#1584,
#2562,
#2947,

This solution will not fix already broken saves where the patient has already reached this dead-end state. (#2562)

But will fix crashes in #1584 and #2947 (but 2947 also have another queue problem) because save made several ticks before a patient trying to start dying scenario.

**Comment by ARGAMX** [comment]
It seems that the problem here is that the patient is "stuck".

```yaml
######## Action Queue. Variant 1 ########
Action: 1 | Name: use_object bench | MH: + | NO_TR: - | NO_INT: -
Action: 2 | Name: queue | MH: + | NO_TR: - | NO_INT: -
Action: 3 | Name: walk | MH: + | NO_TR: - | NO_INT: -
Action: 4 | Name: meander | MH: - | NO_TR: - | NO_INT: -
Action: 5 | Name: die | MH: - | NO_TR: - | NO_INT: -

######## Action Queue. Variant 2 ########
humanoid - class: Standard Male Patient
Warmth: 0.324   Happiness: 1.000   Fatigue: 0.000  Thirst: 0.493  Toilet_Need: 0.528   Health: 0.000   Service Quality: 0.000
Actions: [use_object - bench   must_happen, queue - Bench distance: 0 Standing: false   must_happen, walk - going to 93:68   must_happen, meander, die]
######## ######## ########
```

walk action to the room door for some reason is marked as `must_happen true`.

In exactly the same way as [this patient](https://github.com/CorsixTH/CorsixTH/issues/2947#issuecomment-3145827674) "stuck".

That's why he couldn't get up from the bench to die. 
He had a dying scenario lined up. 
But he can't pop other actions from actions queue.

So this and #2947 is related.
